### ISSUE TITLE
@EnsuresNonNull[If] should ensure all fields on a path are non-null

### ISSUE THREAD

**Issue by wmdietl** [typetools/checker-framework#1917]
In the example below, method `bar` ensures that `this.field.value` is non-null.
However an error for the dereference of `this.field` is generated.
I looked at the Store and it only contains a value for `this.field.value`, not for `this.field`, which explains the error.
Probably similar problems exist for other flow expressions.

```java
import org.checkerframework.checker.nullness.qual.EnsuresNonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

class Demo {
    class A {
        @Nullable Object value;
    }
    
    @Nullable A field;
    
    @EnsuresNonNull("this.field.value")
    void bar() {
        this.field = new A();
        this.field.value = "";
    }
    
  void foo() {
    bar();
    this.field.toString();
  }
}
```

I'm working on an improvement to the CFG & Store and noticed this issue when debugging an issue at
https://github.com/typetools/checker-framework/blob/master/checker/src/org/checkerframework/checker/initialization/InitializationVisitor.java#L183
I don't understand why that code doesn't fail on master, but this false positive illustrates the same issue with the Store.

**Comment by wmdietl** [comment]
Even more confusing:

```java
import org.checkerframework.checker.nullness.qual.EnsuresNonNull;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.checkerframework.checker.nullness.qual.RequiresNonNull;


class Demo {
    class A {
        @Nullable Object value;
    }
    
    @Nullable A field;
    
    @EnsuresNonNull("this.field.value")
    void bar() {
        this.field = new A();
        this.field.value = "";
    }
    
    @RequiresNonNull("this.field.value")
    void baz() {};
    
    void foo() {
      bar();
      baz();
    }
}
```

The call to `baz()` fails with `contracts.precondition.not.satisfied`.

**Comment by smillst** [comment]
```
    @EnsuresNonNull("this.field.value")
    void bar() {
        this.field = new A();
        this.field.value = "";
        this.field = null;
    }
```
